lib/mzgl/midi/AllMidiDevicesAppleImpl.mm

```diff
@@ -335,83 +335,97 @@ static void removeFirst(std::vector<CoreMidiOutRef> &devs, CoreMidiOutRef dev) {
 
 	packet = MIDIPacketListAdd(packetList, sizeof(packetBuffer), packet, 0, size, data);
 
-	sendPacketList(packetList);
+	if (auto endpoint = getEndpointRefForDevice(device, midiOuts); endpoint.has_value()) {
+		auto result = MIDISend(outputPort, *endpoint, packetList);
+		NSLogError(result, "Sending MIDI bytes");
+	} else {
+		NSLog(@"Failed to find end point for device");
+	}
 }
 
-/*
- struct MIDISysexSendRequest
- {
-     MIDIEndpointRef        destination;
-     const Byte *          data;
-     UInt32                bytesToSend;
-     Boolean                complete;
-     Byte                reserved[3];
-     MIDICompletionProc     completionProc;
-     void * __nullable    completionRefCon;
- };
- */
-
 static void cleanupSysex(MIDISysexSendRequest *request) {
-	//    delete [] request->data;
-	//    printf("cleanup %d\n", request->complete);
 	delete request;
 }
 
-void AllMidiDevicesAppleImpl::sendSysex(const UInt8 *data, UInt32 size) {
-	//    NSLog(@"%s(%u sysex bytes to core MIDI)", __func__, unsigned(size));
+void AllMidiDevicesAppleImpl::sendSysex(const MidiDevice &device, const MidiMessage &midiMessage) {
+	const auto *data = midiMessage.getBytes().data();
+	const auto size	 = midiMessage.getBytes().size();
+
+	static constexpr auto maxSize = 655336;
 	assert(size < 65536);
 
-	for (ItemCount index = 0; index < MIDIGetNumberOfDestinations(); ++index) {
-		MIDISysexSendRequest *request = new MIDISysexSendRequest();
+	if (auto endpoint = getEndpointRefForDevice(device, midiOuts); endpoint.has_value()) {
+		auto request = new MIDISysexSendRequest;
 
 		request->data = new Byte[size];
 		memcpy((void *) request->data, (void *) data, size);
+
 		request->bytesToSend	  = size;
 		request->completionProc	  = &cleanupSysex;
 		request->completionRefCon = request;
 		request->complete		  = false;
-		request->destination	  = MIDIGetDestination(index);
-		OSStatus res			  = MIDISendSysex(request);
-		if (res != noErr) {
-			printf("error sending sysex\n");
-		}
+		request->destination	  = *endpoint;
+
+		auto result = MIDISendSysex(request);
+		NSLogError(result, "Sending MIDI sysex");
 	}
 }
 
-void AllMidiDevicesAppleImpl::sendMessage(const MidiMessage &m) {
-	const auto b = m.getBytes();
-	if (b.size() > 0) {
-		if (m.isSysex()) {
-			sendSysex(b.data(), b.size());
-		} else {
-			sendBytes(b.data(), b.size());
-		}
-	} else {
+bool midiMessageHasValidBytes(const MidiMessage &midiMessage) {
+	if (midiMessage.getBytes().empty()) {
 		Log::e() << "Midi message contained no bytes!";
+		return false;
 	}
+
+	return true;
 }
 
-void AllMidiDevicesAppleImpl::sendMessage(const MidiDevice &device, const MidiMessage &m) {
-	// TODO: should only send to the specified device
-	sendMessage(m);
+void AllMidiDevicesAppleImpl::sendMessage(const MidiMessage &midiMessage) {
+	if (!midiMessageHasValidBytes(midiMessage)) {
+		return;
+	}
+
+	for (const auto &output: midiOuts) {
+		sendMessage(*output, midiMessage);
+	}
```

Comment:
To maintain the behaviour previously exhibited, ive changed this so that it loops through all the midi devices and sends the message to it. This should mean that every device gets the message (i.e. its broadcast)